### ISSUE TITLE
Allow a map's key value to be any sequence

### ISSUE THREAD

**Issue by dnovatchev** [qt4cg/qtspecs#119]
Since being introduced in XSLT 3.0 and later in XPath 3.1 the **_[map](https://www.w3.org/TR/xpath-functions-31/#maps-and-arrays)_** datatype has become a powerful and expressive tool for programming in XPath.

At present the value of a key of a map can be **_["an arbitrary atomic value"](https://www.w3.org/TR/xpath-functions-31/#maps-and-arrays)_**, thus a sequence of zero or more than one atomic items cannot be used in a map-key specification.

Besides giving us an almost `1 : 1` correspondence to a `JSON` object (when used with arrays, which themselves can be thought of as maps) maps are useful for expressing the tabular representation of a function that has one argument of type  `xs:anyAtomicType`.

It is not possible using a map to naturally express the tabular representation of a function having two or more (or 0) arguments. While something like this can be done using nested maps as in the example below, this technique is cumbersome and error-prone even when having two arguments, and almost prohibitively difficult when applied to expressing functions with more than 2 arguments.

Here is how we could express one possible tabular form of the function `M**N` (`M` to the power of `N`), where the two arguments are of type `xs:positiveInteger`:

```xq
let $m1 := map {1 : 1, 2 : 2, 3 : 3, 4 : 4, 5 : 5, 6 : 6, 7 : 7, 8 :  8, 9 : 9, 10 :10},
    $m2 := map {1 : 1, 2 : 4, 3 : 9, 4 : 16, 5 : 25, 6 : 36, 7 : 49, 8 : 64, 9 : 81, 10 :100},
    $m3 := map {1 : 1, 2 : 8, 3 : 27, 4 : 64, 5 : 125, 6 : 216, 7 : 343, 8 : 512, 9 : 729, 10 :1000},
    (:  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  . :)  
  
    $M := map{ 1: $m1,  2: $m2,  3: $m3 (: ... :)  }
  return $M(3)(4)
```

Here evaluating `$M(3)(4)` produces the value `4**3` (`4` to the power of `3`), that is: 
`64`

**_This proposal is to expand the allowed value-space for a key of a map to any sequence._**

Thus one will be able to write:

```xq
let $M := map{ (2, 2) : 4, (3, 2) : 9, (: . . . :)
               (2, 3) : 8, (3, 3) : 27 (: . . . :)
}
 return
    $M((2, 3))
```

And evaluating this returns the correct result `8`.

**Possible implementation**:
The implementation is very straightforward: just use in **_[op:same-key](https://www.w3.org/TR/xpath-functions-31/#func-same-key)_** as comparison function an improvement of **[`fn:deep-equal()`](https://www.w3.org/TR/xpath-functions-31/#func-deep-equal)**, which has similar behavior, but never throws errors, is context-free, error-free and transitive. Due to lack of fantasy I called this function **`deep-equal-safe()`** and its description is **[here](https://github.com/dnovatchev/FXSL-XSLT2/blob/master/fn-deep-equal-safe.pdf)**

**Note**:
No new datatypes and no changes to the XDM are necessary in order to implement this proposal.

**Comment by martin-honnen** [comment]
XSLT 3 should have some "infrastructure" for that with the support for composite keys in `xsl:key` and `for-each-group` but I think that is restricted to sequences of atomic values or atomized sequences as the keys. Do Java or .NET or other underlying implementation platforms have classes for efficient Dictionaries/HashMaps based on key sequences?

**Comment by dnovatchev** [comment]
@martin-honnen,

> XSLT 3 should have some "infrastructure" for that with the support for composite keys in `xsl:key` and `for-each-group` but I think that is restricted to sequences of atomic values or atomized sequences as the keys. Do Java or .NET or other underlying implementation platforms have classes for efficient Dictionaries/HashMaps based on key sequences?

Yes, in .NET one can define and use dictionaries, whose keys are themselves collections, for example:

```c#
            var myDict1 = new Dictionary<List<string>, string>();
            var myDict2 = new Dictionary<HashSet<int>, int>();

```
And the **[documentation](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-6.0)** says:

"_[Dictionary<TKey,TValue>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-6.0) requires an equality implementation to determine whether keys are equal. You can specify an implementation of the [IEqualityComparer<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1?view=net-6.0) generic interface by using a constructor that accepts a comparer parameter; if you do not specify an implementation, the default generic equality comparer [EqualityComparer<T>.Default](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.equalitycomparer-1.default?view=net-6.0) is used. If type TKey implements the [System.IEquatable<T>](https://docs.microsoft.com/en-us/dotnet/api/system.iequatable-1?view=net-6.0) generic interface, the default equality comparer uses that implementation._"

We could make a map even more flexible by allowing each instance of a **map** to have its own **`compare-keys()`** function.
Then, what the current proposal says is to make **_`deep-equal()`_** the default value for a map's **`compare-keys()`** function.

**Comment by michaelhkay** [comment]
The deep-equal() function is quite unsuitable as a comparator here. 

The is-same-key() function was designed to have three important properties:

(a) it is error-free
(b) it is context-free
(c) it is transitive and commutative

deep-equal() satisfies none of these conditions.

**Comment by dnovatchev** [comment]
@michaelhkay,
 
> The deep-equal() function is quite unsuitable as a comparator here.
> 
> The is-same-key() function was designed to have three important properties:
> 
> (a) it is error-free (b) it is context-free (c) it is transitive and commutative
> 
> deep-equal() satisfies none of these conditions.
> 
> However, we could certainly support sequences of atomic values as keys.

We need then to evolve `deep-equal()` into `deep-equal-error-free()`, that only differs from the former in not throwing any exceptions but returning `false()` instead.

This is definitely possible, does not require big effort, since most of the existing specification and code of `deep-equal()` will be reused.

The only change in the **[existing specification of `deep-equal()`](https://www.w3.org/TR/xpath-functions-31/#func-deep-equal)** would be to replace:

> "**Error Conditions**
> A type error is raised [[err:FOTY0015](https://www.w3.org/TR/xpath-functions-31/#ERRFOTY0015)] if either input sequence contains a function item that is not a map or array."

With:

**Remark**:

**_"`false()` is returned if either input sequence contains a function item that is not a map or array."_**

**Comment by ChristianGruen** [comment]
The following expression is valid, it returns: `2` and `3`.

```xquery
map {  1: 2, 2: 3 }?( (1, 2) )
```

What would we expect as result for the following query?

```
map {  1: 2, 2: 3, (1, 2): 4 }?( (1, 2) )
```

> Do Java or .NET or other underlying implementation platforms have classes for efficient Dictionaries/HashMaps based on key sequences?

I think that should be doable (as XQuery maps are immutable, I assume that most processors use custom map implementations anyway).

**Comment by dnovatchev** [comment]
@ChristianGruen Yes, the `?` lookup operator will need to be adjusted.

Without digging into this deeper, an obvious solution is to have:

```xq
map {  1: 2, 2: 3, (1, 2): 4 }?[1, 2, (1, 2) ]
```

**Comment by michaelhkay** [comment]
@dnovatchev No, there are many more problems with deep-equal than this. There are more errors it can throw, its result depends on the default collation and the implicit timezone, and it's sensitive to the non-transitivity that arises from conversions between numeric types. Very careful thought went into the design of op:same-key to avoid these problems.

**Comment by dnovatchev** [comment]
> @dnovatchev No, there are many more problems with deep-equal than this. There are more errors it can throw, its result depends on the default collation and the implicit timezone, and it's sensitive to the non-transitivity that arises from conversions between numeric types. Very careful thought went into the design of op:same-key to avoid these problems.

@michaelhkay Yes, I see in the existing Xpath 3.1 specification that in two of the four cases of **`op:same-key`** it already delegates to **`deep-equal`**.  We will get the desired new `op:same-key` if we follow the same approach -- keep the same treatment of string and date-time comparisons, but for the other cases do what **`deep-equal`** is doing.

Do you see any additional pitfalls with this?

**Comment by michaelhkay** [comment]
If we're doing this at all, I would

(a) confine the key to sequences of atomic values

(b) extend op:same-key to handle sequences of atomic values (not difficult: the sequences must be the same length and the items must satisfy op:same-key pairwise)

(c) require explicit use of map:get() or function call syntax (not a lookup operator) when using a composite key to retrieve from a map

There are of course myriad other details to be sorted out, for example changed rules for JSON serialization.

Generally though I would like to restate my position that I would prefer in this round to make no data model changes. Data model changes are usually far more disruptive than people imagine, and often make it hard to retain backward compatibility.

**Comment by dnovatchev** [comment]
> If we're doing this at all, I would
> 
> (a) confine the key to sequences of atomic values
> 
> (b) extend op:same-key to handle sequences of atomic values (not difficult: the sequences must be the same length and the items must satisfy op:same-key pairwise)
> 
> (c) require explicit use of map:get() or function call syntax (not a lookup operator) when using a composite key to retrieve from a map

@michaelhkay Generally agree.

But (a) is unnecessary limitation. Why .NET allows us to have:

```c#
            var myDict1 = new Dictionary<List<string>, string>();
            var myDict2 = new Dictionary<HashSet<int>, int>();
```

but we want to avoid such uses for a **`map`**?

Now is the time to do good things, let us do them!

Especially when we already have a standard function in XPath 3.1 that compares any two sequences for equality? It seems a rather arbitrary decision when we allow two sequences to be compared and when not. No logic, and quite confusing!!!

> Generally though I would like to restate my position that I would prefer in this round 

I would agree, if "this round" was confined to a specified period, like 6 months or one year. However we will soon mark the 2-year anniversary of the 1st still open issue, last changed on Nov. 27th 2020.

I completely agree with @ChristianGruen asking several times when we are going to bring "this round" to a close.

In C# - time 2 years is the time for two complete new versions of the language. Why are we so lagging... ?

If something bad happens (strongly hope not!!!) to one of the major contributors, what will happen to their proposals?

And if "this round" is unpredictably long, then why bother at all produce any proposals? Will we at all live to see any of these advance?

Thanks,
Dimitre

**Comment by ChristianGruen** [comment]
> I completely agree with @ChristianGruen asking several times when we are going to bring "this round" to a close.

Thanks for picking this up. So maybe it's advisable not to provide new proposals anymore (how good they may be), but finalize or close the existing ones first.

> In C# - time 2 years is the time for two complete new versions of the language. Why are we so lagging... ?

I’m pretty sure that C# is driven by both more financial and human resources.

**Comment by michaelhkay** [comment]
I'll concentrate on the technical question and not the planning/administrative one.

C# and Java allow you to use any value as a dictionary key, provided that the relevant class has "well behaved" hashCode() and equals() methods. I'm reluctant to have user-defined callbacks with a contract that we can't validate or enforce (e.g. that if two values are equal, then they have the same hash code).

There are very good reasons why we didn't use deep-equal for comparing keys in maps, and I thought I had explained them: we wanted a function that was error-free, context-free, and transitive. There were several hours of discussion on this in the WGs when we did the 3.0 specs, and I really don't want to go through that again.

**Comment by dnovatchev** [comment]
> There are very good reasons why we didn't use deep-equal for comparing keys in maps, and I thought I had explained them: we wanted a function that was error-free, context-free, and transitive. There were several hours of discussion on this in the WGs when we did the 3.0 specs, and I really don't want to go through that again.

@michaelhkay I have a new function: **_`deep-equal-good()`_** which is similar to [**`fn:deep-equal`**](https://www.w3.org/TR/xpath-functions-31/#func-deep-equal) but doesn't raise errors, is transitive, deterministic, context-independent and focus-independent.

I am proposing that the new comparison operator for keys of maps (the "new `op:has-same-key()`" ) uses exclusively `deep-equal-good()` for comparing keys.

The description of the function is **_[here](https://github.com/dnovatchev/FXSL-XSLT2/blob/master/fn-deep-equal-good.pdf)_**

Any comments / observations on this proposed comparison function are welcome.

**Comment by rhdunn** [comment]
For `deep-equal`, we could take advantage of my https://github.com/qt4cg/qtspecs/issues/98 proposal and add flags to support this behaviour.

**Comment by ChristianGruen** [comment]
I can see there may be use cases for maps with sequences as keys, but I wonder if this needs to be available in the core language. I would rather love to have extension modules for advanced data structures, in the EXPath domain or somewhere else (such as https://github.com/LeoWoerteler/xq-modules), as there are numerous other data structures we could envision.

**Comment by michaelhkay** [comment]
Agreed. I want to resist doing things that involve data model changes if at all possible. I think it's possible to define a "composite key map" (as a map of maps) using the existing data model, with supporting functions such as `mm:get()`, `mm:put()`, `mm:contains()`, `mm:for-each()`, etc, and this would avoid the disruption of having to revise things such as the lookup operator, the JSON serialization method, the deep-equal() function etc. Data model changes tend to have very pervasive effects.

**Comment by dnovatchev** [comment]
@michaelhkay 

>  the disruption of having to revise things such as the lookup operator, the JSON serialization method, the deep-equal() 
> function etc. Data model changes tend to have very pervasive effects.

We will not revise the lookup operator -- for a non-atomic key we will only allow 

`$map($non-atomic-key)`

As for revising `deep-equal()` this proposal simply advocates using a new function, not revising `deep-equal()`. I see that  there is already a proposal to add options for the behavior of `deep-equal()` and if this is accepted we can even live with using `deep-equal()` with options to behave as `op:same-key()` when comparing atomic values.

And finally, with the proposal implemented, the JSON serialization of maps with atomic-keys will remain unchanged.

These are answers to all above-stated plausible issues.

Do you have more in mind?

Thanks,
Dimitre

**Comment by benibela** [comment]
Rather than allowing sequences as keys, it might be more reasonable to allow arrays as keys. That might be less disruptive since arrays have  count 1 like atomic values

**Comment by michaelhkay** [comment]
For this kind of proposal, I think it's useful to do a comprehensive document listing in detail, section by section, the textual changes needed to each one of the specs. There's an immense amount of detail needed. To take just one small example, the map:keys() function doesn't extend in any natural way. To take another, how do you propose to define map types? Do you just extend `map(item-type, sequence-type)` to `map(sequence-type, sequence-type)`? But that would imply the sequence of keys is homogenous. Rather than discussing each of these details individually, I think a thorough and detailed design paper is needed.

**Comment by dnovatchev** [comment]
> Rather than allowing sequences as keys, it might be more reasonable to allow arrays as keys. That might be less disruptive since arrays have count 1 like atomic values

@benibela Really like this idea. An array is a single item and it is perfect as a container for any (even non-flat!) sequence of keys. Using an array instead of a sequence of key values seems not to be susceptible at least to some of the issues mentioned by @michaelhkay .

The proposal already includes using a map or an array as a key value. Thus if a sequence `$seq` is meant to be a key-value for a given map, the key-value can be specified as `[$seq]`

**Comment by ndw** [comment]
Apologies, I must have got the number wrong in that PR.

**Comment by Arithmeticus** [comment]
@dnovatchev @benibela I would want to hear more about opening the definition of map key to include an array. That would seem to me difficult or impossible, because arrays can include maps, arrays, and functions as members, recursively, and I think that would pose problems. Unless one allowed only arrays with sequences of atomic items. Not sure how this would look.

Also, under this proposal, what are the implications for the signature of `map:keys()`? Would it return a sequence of arrays? An array of sequences? Something else? It seems to me that the elegance of that function could be compromised.

**Comment by dnovatchev** [comment]
@Arithmeticus Thanks for your questions!

> @dnovatchev @benibela I would want to hear more about opening the definition of map key to include an array. That would seem to me difficult or impossible, because arrays can include maps, arrays, and functions as members, recursively, and I think that would pose problems. Unless one allowed only arrays with sequences of atomic items. Not sure how this would look.

Such recursive comparison is already described for **_[`fn:deep-equal`](https://www.w3.org/TR/xpath-functions-31/#func-deep-equal)_** in XPath 3.1 (See the text starting with: **_"If $i1 and $i2 are both [·arrays·](https://www.w3.org/TR/xpath-functions-31/#),"_**) and no problems have been reported in its implementations by various vendors.

> 
> Also, under this proposal, what are the implications for the signature of `map:keys()`? Would it return a sequence of arrays? An array of sequences? Something else? It seems to me that the elegance of that function could be compromised.

Good question.

This will be a sequence of atomic values (each corresponding to an atomic-value key) and arrays (each corresponding to a composite key (sequence, array or map). Hope this can be expressed well as a union of types.

Thus, for existing, pre-4.0 maps, **_`map:keys()`_** will continue to produce exactly the same result. 

Some of the benefits of using an array for specifying a composite (or an empty-sequence) keys:

- Uniform representation of a non-atomic key -- always as an array
- Ability to contain nested (non-flat sequences)
- Ability to contain and preserve the empty sequence `()` thus allowing a map to represent a function that can have the empty sequence as an argument, which is not allowed under the current XPath 3.1 Maps Specification.

![image](https://user-images.githubusercontent.com/10605892/197922917-f57c622f-e19e-47ac-8d59-9185be3fd550.png)

**Comment by michaelhkay** [comment]
>For this kind of proposal, I think it's useful to do a comprehensive document listing in detail, section by section, the textual changes needed to each one of the specs. 

I started doing this, and got quite a long way, but then hit a significant stumbling block.

I was assuming that we could simply extend functions like `map:put`, `map:get`, `map:contains`, and `map:remove` so the second argument is of type `item()*` rather than `xsanyAtomicType`. However, this breaks compatibility, because at present if you say `map:get($map, @code)`, the second argument is atomised: it takes the value of the `@code` attribute as the required key, not the node identity of the attribute node.

Possible solutions to this might be:

* a whole new raft of functions, perhaps in a new namespace, for "generalized maps"
* options on the current functions to say, in some way, "don't atomize the key"
* a data model change so that "generalized maps" are distinguished from "atomic" maps by some kind of property, and the decision whether or not to atomize the supplied key is based on what kind of map we are dealing with

The suggestion of accepting arrays as key values doesn't solve the problem, because arrays are currently atomized.

These options all feel to me highly unattractive in terms of their impact on the complexity of the specifications. Unless we can find an elegant solution to this difficulty, I conclude that this added complexity is too high a price to pay for what is only a "nice to have" feature. 

Furthermore, if we do find a way of enhancing maps to accept nodes, then comparing nodes by identity will meet some users' needs but not others.

And: there's already a simple workaround to creating a map based on node identity: just use generate-id() for the key.

Note that these objections don't apply if we merely generalise maps to allow a sequence of atomic values as the key. That can be done in a reasonably compatible way; the only difficulty being that the `map:keys` function no longer delivers a useful result, which is easily fixed by providing an alternative function that returns the keys as an array.

**Comment by ChristianGruen** [comment]
We also need to think of the lookup operator. Currently, these two expressions return a value…

```xquery
map { <இ>a</இ>: 1 }?a
map { 'a': 1 }?(<இ>a</இ>)
```

…and the behavior shouldn’t change with version 4.

**Comment by michaelhkay** [comment]
There might be a solution that follows the precedent of `fn:collation-key`: define a function `map:identity-key()` that takes an arbitrary sequence as input and returns an implementation-defined binary value that encapsulates the supplied sequence. You could then do `map:put($map, map:identity-key(@code))` and `map:get($map, map:identity-key(@code))`.

**Comment by dnovatchev** [comment]
> I was assuming that we could simply extend functions like `map:put`, `map:get`, `map:contains`, and `map:remove` so the second argument is of type `item()*` rather than `xsanyAtomicType`. However, this breaks compatibility, because at present if you say `map:get($map, @code)`, the second argument is atomised: it takes the value of the `@code` attribute as the required key, not the node identity of the attribute node.
> 
> Possible solutions to this might be:
> 
> * a whole new raft of functions, perhaps in a new namespace, for "generalized maps"
> * options on the current functions to say, in some way, "don't atomize the key"
> * a data model change so that "generalized maps" are distinguished from "atomic" maps by some kind of property, and the decision whether or not to atomize the supplied key is based on what kind of map we are dealing with
> 
> The suggestion of accepting arrays as key values doesn't solve the problem, because arrays are currently atomized.
> 
> These options all feel to me highly unattractive in terms of their impact on the complexity of the specifications. Unless we can find an elegant solution to this difficulty, I conclude that this added complexity is too high a price to pay for what is only a "nice to have" feature.

There is a simpler solution. Just have the new functions:

**map:put-with-node($map as map(\*), $key as node(), $value as item()\*) as map(\*)**

and

**map:get-from-node($map as map(\*), $key as node()) as item()\***

This may need further elaboration, but I think it is doable.

**Comment by ChristianGruen** [comment]
I propose to close this issue. We could create a separate issue on storing non-atomic items, such as nodes, in maps (the last comments of the discussion seemed to have focused on this aspect).

**Comment by dnovatchev** [comment]
> I propose to close this issue. We could create a separate issue on storing non-atomic items, such as nodes, in maps (the last comments of the discussion seemed to have focused on this aspect).

@ChristianGruen 

This issue is not about "storing non-atomic items, such as nodes, in maps" which can be done smoothly in XPath 3.

The issue is about using non-atomic items as **_keys_** of a map.

And I think @michaelhkay  has done already impressive work on this (judging from several related issues). Just to quote from one of his latest comments in this thread:

> Note that these objections don't apply if we merely generalise maps to allow a sequence of atomic values as the key. That can be done in a reasonably compatible way; the only difficulty being that the `map:keys` function no longer delivers a useful result, which is easily fixed by providing an alternative function that returns the keys as an array.

And:

> There might be a solution that follows the precedent of `fn:collation-key`: define a function `map:identity-key()` that takes an arbitrary sequence as input and returns an implementation-defined binary value that encapsulates the supplied sequence. You could then do `map:put($map, map:identity-key(@code))` and `map:get($map, map:identity-key(@code))`.

Why bury the results of all this work exactly when it seems we are close to a solution?

**Comment by ChristianGruen** [comment]
> The issue is about using non-atomic items as **_keys_** of a map.

Yes, that’s what I wanted to express.

> Why bury the results of all this work exactly when it seems we are close to a solution?

Once again no reason to take it personal, Dimitre. The last comment was 9 months ago; I’ll be fine if we manage to move things forward.

For example, I’d still be interested how we believe we can handle lookups. See my last comment in this discussion, for which I, haven't got a reply so far.

**Comment by dnovatchev** [comment]
> > The issue is about using non-atomic items as **_keys_** of a map.
> 
> Yes, that’s what I wanted to express.
> 
> > Why bury the results of all this work exactly when it seems we are close to a solution?
> 
> Once again no reason to take it personal, Dimitre. The last comment was 9 months ago; I’ll be fine if we manage to move things forward.
> 
> For example, I’d still be interested how we believe we can handle lookups. See my last comment in this discussion, for which I, haven't got a reply so far.

We have done so many things, but I vaguely remember that @michaelhkay did propose an error-free comparison function - not sure about its name.

We need to revisit the current proposal in light of such latest achievements - I believe we are moving closer to being able to have this feature.

Please, correct me if this is not so.

@michaelhkay  ?

**Comment by ChristianGruen** [comment]
> We have done so many things, but I vaguely remember that @michaelhkay did propose an error-free comparison function - not sure about its name.

@dnovatchev To understand this better, how would the “comparison function” be combined with the lookup operator? Did you think of `map:identity-key`?

**Comment by michaelhkay** [comment]
I'm going to propose closing this with no action, since we had lengthy discussion to explore the possible complications and possible solutions, and this did not lead to a concrete proposal. Of course, closing the issue doesn't prevent anyone raising  a concrete proposal; rather it signals merely that the CG feels it would be acceptable to publish a 4.0 specification without this feature.

**Comment by ndw** [comment]
The CG agreed to close this issue without further action at [meeting 070](https://qt4cg.org/meeting/minutes/2024/03-19.html#close-without-action).

**Comment by dnovatchev** [comment]
> Since being introduced in XSLT 3.0 and later in XPath 3.1 the **_[map](https://www.w3.org/TR/xpath-functions-31/#maps-and-arrays)_** datatype has become a powerful and expressive tool for programming in XPath.
> 
> At present the value of a key of a map can be **_["an arbitrary atomic value"](https://www.w3.org/TR/xpath-functions-31/#maps-and-arrays)_**, thus a sequence of zero or more than one atomic items cannot be used in a map-key specification.
> 
> Besides giving us an almost `1 : 1` correspondence to a `JSON` object (when used with arrays, which themselves can be thought of as maps) maps are useful for expressing the tabular representation of a function that has one argument of type `xs:anyAtomicType`.
> 
> It is not possible using a map to naturally express the tabular representation of a function having two or more (or 0) arguments. While something like this can be done using nested maps as in the example below, this technique is cumbersome and error-prone even when having two arguments, and almost prohibitively difficult when applied to expressing functions with more than 2 arguments.
> 
> Here is how we could express one possible tabular form of the function `M**N` (`M` to the power of `N`), where the two arguments are of type `xs:positiveInteger`:
> 
> let $m1 := map {1 : 1, 2 : 2, 3 : 3, 4 : 4, 5 : 5, 6 : 6, 7 : 7, 8 :  8, 9 : 9, 10 :10},
>     $m2 := map {1 : 1, 2 : 4, 3 : 9, 4 : 16, 5 : 25, 6 : 36, 7 : 49, 8 : 64, 9 : 81, 10 :100},
>     $m3 := map {1 : 1, 2 : 8, 3 : 27, 4 : 64, 5 : 125, 6 : 216, 7 : 343, 8 : 512, 9 : 729, 10 :1000},
>     (:  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  . :)  
>   
>     $M := map{ 1: $m1,  2: $m2,  3: $m3 (: ... :)  }
>   return $M(3)(4)
> Here evaluating `$M(3)(4)` produces the value `4**3` (`4` to the power of `3`), that is: `64`
> 
> **_This proposal is to expand the allowed value-space for a key of a map to any sequence._**
> 
> Thus one will be able to write:
> 
> let $M := map{ (2, 2) : 4, (3, 2) : 9, (: . . . :)
>                (2, 3) : 8, (3, 3) : 27 (: . . . :)
> }
>  return
>     $M((2, 3))
> And evaluating this returns the correct result `8`.
> 
> **Possible implementation**: The implementation is very straightforward: just use in **_[op:same-key](https://www.w3.org/TR/xpath-functions-31/#func-same-key)_** as comparison function an improvement of **[`fn:deep-equal()`](https://www.w3.org/TR/xpath-functions-31/#func-deep-equal)**, which has similar behavior, but never throws errors, is context-free, error-free and transitive. Due to lack of fantasy I called this function **`deep-equal-safe()`** and its description is **[here](https://github.com/dnovatchev/FXSL-XSLT2/blob/master/fn-deep-equal-safe.pdf)**
> 
> **Note**: No new datatypes and no changes to the XDM are necessary in order to implement this proposal.

**Comment by dnovatchev** [comment]
I have reopened this issue, because:

1. I am convinced in the benefits to be achieved.
2. We were close to an agreement - see some of the comments by @michaelhkay.
3. We already are able (even in XPath 3.1) to successfully compare, even sort, non-atomic items, such as maps, arrays and sequences

**Comment by michaelhkay** [comment]
Allowing the key to be any *atomic* sequence is certainly do-able. It does, of course, involve a change to the data model, and there are non-trivial decisions to be made for example about what the map:keys() function should return. Also about the type system: what happens to the type map(K, V); can K be a heterogeous type such as (int, string); do all the keys have to be sequences of the same length. 

So the fact that it's do-able doesn't mean it's trivial. There will be many detailed questions that need to be thrashed out.

**Comment by dnovatchev** [comment]
> So the fact that it's do-able doesn't mean it's trivial. There will be many detailed questions that need to be thrashed out.

Absolutely!

Let us do our job. If we changed the DM on much more trivial  occasions, as for example "ordered maps", let us do this in honest for something really important.

**Comment by michaelhkay** [comment]
My advice would be: take this one step at a time. The bigger the proposal, the more obstacles you hit getting it past the review process.

**Comment by ndw** [comment]
The CG agreed to merge close this issue without further action at [meeting 110](https://qt4cg.org/meeting/minutes/2025/02-18.html).

**Comment by dnovatchev** [comment]
I don't remember any mention of closing the issue at today's meeting.

On the contrary, this issue was mentioned as one of the good ways to provide relaxed bounds/key checking of arrays/maps.

I am reopening the issue, as the closing was done either by mistake, or as result of a (not discussed) bias towards some solutions to the relaxed bounds checking at the expense of others.

Actually, at today's meeting it was stated that the current issue might be a good solution to the relaxed bounds checking problem, and this only raises its importance.
